### PR TITLE
Sharded deploy v2 agent version

### PR DESCRIPTION
**Pull request by ankilosaurus**
Ready to merge.

Related deploy-agent changes in PR: #746

### CODE REVIEW

File: tools/mysql/schema-update-8.sql [hunk 1]
```diff
@@ -0,0 +1,10 @@
+-- !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+-- ALWAYS BACKUP YOUR DATA BEFORE EXECUTING THIS SCRIPT
+-- !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
```
Comment (ankilosaurus) [inline tools/mysql/schema-update-8.sql:3]: combine into one schema change file

File: deploy-agent/deployd/agent.py [hunk 2]
```diff
@@ -388,6 +388,14 @@ def main():
     parser.add_argument('--script-variables', dest='script_variables', default='{}',
                         help="Optional. In 'serverless' mode,  script_variables is needed in "
                              "json format.")
+    parser.add_argument('--availability-zone', dest="availability_zone", required=False, default=None,
+                        help="Availability zone being used when interact with Teletraan service. "
+                             "This is optional. By default the availability zone defined in host-info "
+                             "file will be used")
```
Comment (ankilosaurus) [inline deploy-agent/deployd/agent.py:394]: not needed.

File: deploy-service/common/src/main/java/com/pinterest/deployservice/handler/PingHandler.java [hunk 3]
```diff
@@ -435,6 +441,31 @@ PingRequestBean normalizePingRequest(PingRequestBean pingRequest) throws Excepti
         return pingRequest;
     }
 
+    Set<String> shardGroups(PingRequestBean pingRequest) throws Exception {
+        Set<String> shards = new HashSet<>();
+
+        String asg = pingRequest.getAutoscalingGroup();
+        EnvironBean envBean = environDAO.getByCluster(asg);
```
Comment (ankilosaurus) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/handler/PingHandler.java:448]: null check on asg?

File: deploy-agent/deployd/client/client.py [hunk 4]
```diff
@@ -54,6 +57,13 @@ def _read_host_info(self):
                 output = utils.get_info_from_facter(self._config.get_facter_group_key())
                 if output:
                     self._hostgroup = output.split(",")
+
+            if not self._availability_zone:
+                self._availability_zone = utils.get_info_from_facter("ec2_metadata.placement.availability-zone")
+
+            # Hosts brought up outside of ASG or Teletraan might not have ASG
+            if not self._autoscaling_group:
+                self._autoscaling_group = utils.get_ec2_tag_from_file(self._config.get_ec2_tags_cache(), "Autoscaling")
```
Comment (ankilosaurus) [inline deploy-agent/deployd/client/client.py:66]: check with Jayme

**Comment by yongwen** [comment]
1. separate deployagent py into different PR
2. check the get_ec2_tag_from cache impl
3. separate env url validation from other PR
4. sql host_name -> host_id
5. deployagent.py no additional command line args

**Comment by ankilosaurus** [comment]
deploy agent changes in separate PR #746

**Pull request merged**